mempool/src/mempool.rs

```diff
@@ -511,11 +619,77 @@ impl Mempool {
 
         debug!(
             returned_txs = tx_vec.len(),
-            remaining_txs = mempool_state_upgraded.transactions.len(),
-            "Returned transactions from mempool"
+            remaining_txs = mempool_state_upgraded
+                .regular_transactions
+                .transactions
+                .len(),
+            "Returned regular transactions from mempool"
         );
 
-        tx_vec
+        (tx_vec, size)
+    }
+
+    /// Returns a vector with accepted control transactions from the mempool.
+    ///
+    /// Returns the highest fee per byte up to max_bytes transactions and removes them from the mempool
+    pub fn get_control_transactions_for_block(
```

Comment:
I moved the logic for the expired transactions into the `MempoolState`. I did not do it for `get_transactions_for_block` because I am unsure wether or not more fine grained control over the byte size in particular is required.
However I believe even then a function like:
```
impl MempoolState {
    pub fn get_transactions_for_block(&self, max_bytes: (usize, usize)) -> Vec<Transactions>) {
        ...
    }
}
```
might be the way to go here. One thing in particular that I see is that for both the regular and the control transactions a `self.state.upgradable_read()` and a subsequent upgrade of the lock are required. Moving the logic to the `MempoolState` would remove that. The code is also pretty much identical, aside from which field to access. Abstracting what transactions shall be retrieved might also be a good idea.

If that is not desirable, I probably would still move both these functions in the more lightweight `MempoolState` and just call them from here after taking the lock.